lib/Dialect/FIRRTL/Transforms/LowerXMR.cpp

```diff
@@ -241,23 +270,92 @@ class LowerXMRPass : public LowerXMRBase<LowerXMRPass> {
                            });
   }
 
+  void markForRemoval(Operation *op) { opsToRemove.push_back(op); }
+
+  Optional<size_t> getRemoteRefSend(Value val) {
+    auto iter = dataflowAt.find(dataFlowClasses.getOrInsertLeaderValue(val));
+    if (iter != dataflowAt.end())
+      return iter->getSecond();
+    // The referenced module must have already been analyzed, error out if the
+    // dataflow at the child module is not resolved.
+    if (BlockArgument arg = val.dyn_cast<BlockArgument>())
+      arg.getOwner()->getParentOp()->emitError(
+          "reference dataflow cannot be traced back to the remote read op "
+          "for module port '")
+          << dyn_cast<FModuleOp>(arg.getOwner()->getParentOp())
+                 .getPortName(arg.getArgNumber())
+          << "'";
+    else
+      val.getDefiningOp()->emitOpError(
+          "reference dataflow cannot be traced back to the remote read op");
+    signalPassFailure();
+    return None;
+  }
+
+  size_t addReachingSendsEntry(Value atRefVal, Attribute newRef,
+                               Optional<size_t> continueFrom = None) {
+    auto leader = dataFlowClasses.getOrInsertLeaderValue(atRefVal);
+    auto indx = refSendPathList.size();
+    dataflowAt[leader] = indx;
+    refSendPathList.push_back(std::make_pair(newRef, continueFrom));
+    return indx;
+  }
+
+  void garbageCollect() {
+    // Now erase all the Ops and ports of RefType.
+    // This needs to be done as the last step to ensure uses are erased before
+    // the def is erased.
+    for (Operation *op : llvm::reverse(opsToRemove))
+      op->erase();
+    for (auto iter : refPortsToRemoveMap)
+      if (auto mod = dyn_cast<FModuleOp>(iter.getFirst()))
+        mod.erasePorts(iter.getSecond());
+      else if (auto inst = dyn_cast<InstanceOp>(iter.getFirst())) {
+        ImplicitLocOpBuilder b(inst.getLoc(), inst);
+        inst.erasePorts(b, iter.getSecond());
+        inst.erase();
+      }
+    opsToRemove.clear();
+    refPortsToRemoveMap.clear();
+    dataflowAt.clear();
+    refSendPathList.clear();
+  }
+
   /// Cached module namespaces.
   DenseMap<Operation *, ModuleNamespace> moduleNamespaces;
 
-  /// Map of a reference value to the path to the RefSend op. The path to
-  /// RefSend is represented as an ArrayAttr, which is an array of the InnerRef
-  /// to InstanceOps. The path is required since there can be multiple paths to
-  /// the RefSend and we need to identify a unique path. Each ref value can be
-  /// satically resolved to a single remote send op, according to the
-  /// constraints on the RefType. The ArrayAttr ensures that only unique copies
-  /// of the path exist.
-  DenseMap<Value, ArrayAttr> reachingRefSendAt;
+  DenseSet<Operation *> visitedModules;
+  /// Map of a reference value to an entry into refSendPathList. Each entry in
+  /// refSendPathList represents the path to RefSend. i
```

Comment:
```suggestion
  /// refSendPathList represents the path to RefSend.
```
It may be worth mentioning this maps from EC leaders.